### PR TITLE
Bug 1211501 - Support Marionette protocol level 3; r=gaye

### PR DESCRIPTION
**Pull request by andreastt**

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=af7c8e7b73ec3295b653b3ca6af34118a1d49f8e

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=16188ef2afece218e5a4dd39464b6e957266719b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=772f4220b5bbc614cc84618e4aaff68f1bf5567d

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=941654494622ac649e4900511b2478db8077c865

File: tests/jsmarionette/client/marionette-client/lib/marionette/client.js [hunk 1]
```diff
@@ -468,20 +469,27 @@
         this._bypassScopeChecks = false;
       }
 
+      var data;
       if (this.protocol == 1) {
-        if (!cmd.to) {
-          cmd.to = this.actor || 'root';
+        if (!body.to) {
+          body.to = this.actor || 'root';
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/client.js]: Also could be written `body.to = body.to || this.actor || 'root';` but doesn't matter a ton :)

File: tests/jsmarionette/client/marionette-client/lib/marionette/client.js [hunk 2]
```diff
@@ -468,20 +469,27 @@
...
         }
         if (this.sessionId) {
-          cmd.session = cmd.session || this.sessionId;
+          body.session = body.session || this.sessionId;
         }
+        data = body;
+      } else if (this.protocol == 3) {
+        var cmd = new message.Command(body.name, body.parameters);
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/client.js]: nit: Do we need this `cmd` var?

Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/client.js]: Also can we just send the `body` to the `message.Command` constructor?

File: tests/jsmarionette/client/marionette-client/lib/marionette/drivers/abstract.js [hunk 3]
```diff
@@ -46,29 +47,29 @@
...
      * @property connectionId
-     * @type Numeric
+     * @type Number
      */
     connectionId: null,
 
     /**
-     * We just set the script timeout.
-     * If you need to do something in the driver.
+     * Duration in milliseconds defining when to abort an attempt to
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/drivers/abstract.js]: I appreciate this!

File: tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js [hunk 4]
```diff
@@ -168,13 +164,71 @@ TcpSync.prototype._readResponse = function() {
...
+  var res;
+
+  if (this.marionetteProtocol >= 3) {
+    var resp = Response.fromMsg(data);
+    res = resp.error || resp.result;
+  } else {
+    res = data;
+  }
+
+  return res;
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js]: If all you're doing is returning `res` you could just return above and avoid nesting

File: tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js [hunk 5]
```diff
@@ -168,13 +164,71 @@ TcpSync.prototype._readResponse = function() {
...
+ *     The return value from the passed in callback.
+ */
+TcpSync.prototype.send = function(obj, cb) {
+  var send;
+  if (obj instanceof Command || obj instanceof Response) {
+    send = this.sendMessage;
+  } else {
+    send = this.sendRaw;
+  }
+  return send.bind(this)(obj, cb);
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js]: `this.send(obj, cb);`?

File: tests/jsmarionette/client/marionette-client/lib/marionette/message.js [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+'use strict';
+
+function Command(name, params) {
+  this.id = null;
+  this.name = name;
+  this.params = params || {};
+}
+
+Command.prototype.toString = function() {
+  return 'Command {id: ' + this.id + ', ' +
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/message.js]: Perhaps cleaner if you do `'Command ' + JSON.stringify({ ... })`?

File: tests/jsmarionette/client/marionette-client/lib/marionette/message.js [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+  if (obj instanceof Error) {
+    this.error = obj;
+  } else {
+    this.result = obj;
+  }
+}
+
+Response.prototype.toString = function() {
+  return 'Response {id: ' + this.id + ', ' +
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/message.js]: ditto

File: tests/jsmarionette/client/marionette-client/lib/marionette/message.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+};
+
+Response.fromMsg = function(msg) {
+  return new Response(msg[1], msg[2] || msg[3]);
+};
+
+Response.TYPE = 1;
+
+module.exports = {
+  'Command': Command,
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/message.js]: No need for quotes

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=3c27cb24e1c3ad3555c661165602699595de8264

File: tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js (hunk 5 above)
Comment (andreastt) [inline tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js]: I tried being a bit too clever my own good here.

File: tests/jsmarionette/client/marionette-client/lib/marionette/message.js (hunk 6 above)
Comment (andreastt) [inline tests/jsmarionette/client/marionette-client/lib/marionette/message.js]: Oh, very clever.  Thanks!

File: tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js (hunk 4 above)
Comment (andreastt) [inline tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js]: That makes sense.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=668cf3fd19f0e873a552b503d494df47eed49be6

**Comment by andreastt** [comment]
@gaye Thanks for the valuable feedback!

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=668cf3fd19f0e873a552b503d494df47eed49be6

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2be30e50c170846dca283b7b1033296fbfcd53a0

File: tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js [hunk 9]
```diff
@@ -168,13 +164,65 @@ TcpSync.prototype._readResponse = function() {
     throw error;
   }
 
-  return data;
+  if (this.marionetteProtocol >= 3) {
+    var resp = Response.fromMsg(data);
+    return resp.error || resp.result;
+  } else {
```
Comment (gaye) [inline tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js]: No need for `else` :)

**Comment by gaye** [comment]
It seems like Gij doesn't like your changeset

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=77e3391f8e538c2120486f6997a381a92d353f7e

File: tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js (hunk 9 above)
Comment (andreastt) [inline tests/jsmarionette/client/marionette-client/lib/marionette/drivers/tcp-sync.js]: Thanks, corrected!

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=21fffc62c6a6f1ebacab635776881a518e96b6a5

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8ac880e40f17a7f1f925036db68381ac783b87a1

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=98f5572e5f4afa02abe8791403a28048a0789a84

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=aa95b95f62e07d99ccd5b6e00beb3b9b831bacd5

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=67d3f74f49bfadaaf4b284909a36e8bc79d552d0

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=78f56c00edeba90897f85a4ade06cb86026bcda2

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8239a716a17223f04bc641eaedb7c417f9562cef

**Pull request merged**